src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -2850,3 +2850,146 @@ bool LibraryCallKit::inline_vector_compress_expand() {
+  int vadd_op = VectorNode::opcode(add_op, elem_bt);
+  int vmul_op = VectorNode::opcode(mul_op, elem_bt);
+
+  // Check whether the basic ops are supported by the current hardware
+  if (!arch_supports_vector(Op_VectorLoadConst, num_elem, iota_bt, VecMaskNotUsed) ||
+      !arch_supports_vector(vmul_op, num_elem, elem_bt, VecMaskNotUsed) ||
+      !arch_supports_vector(vadd_op, num_elem, elem_bt, VecMaskNotUsed)) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** not supported: vlen=%d etype=%s", num_elem, type2name(elem_bt));
+    }
+    return false; // not supported
+  }
+
+  // If it is a floating point type vector, additionally check whether the
+  // relative vector cast op is supported by the current hardware
+  if (is_floating_point_type(elem_bt)) {
+    int vcast_op = elem_bt == T_FLOAT ? Op_VectorCastI2X : Op_VectorCastL2X;
+    if (!arch_supports_vector(vcast_op, num_elem, elem_bt, VecMaskNotUsed)) {
+      if (C->print_intrinsics()) {
+        tty->print_cr("  ** not supported: vlen=%d etype=%s", num_elem, type2name(elem_bt));
+      }
+      return false; // not supported
+    }
+  }
+
+  ciKlass* vbox_klass = vector_klass->const_oop()->as_instance()->java_lang_Class_klass();
+  const TypeInstPtr* vbox_type = TypeInstPtr::make_exact(TypePtr::NotNull, vbox_klass);
+  Node* opd = unbox_vector(argument(3), vbox_type, elem_bt, num_elem);
+  if (opd == NULL) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** unbox failed vector=%s",
+                    NodeClassNames[argument(3)->Opcode()]);
+    }
+    return false;
+  }
+
+  // Compute the iota indice vector
+  const TypeVect* vt = TypeVect::make(elem_bt, num_elem);
+  const TypeVect* iota_vt = TypeVect::make(iota_bt, num_elem);
+  Node* iota = gvn().transform(new VectorLoadConstNode(gvn().makecon(TypeInt::ZERO), iota_vt));
+  if (elem_bt == T_FLOAT) {
+    iota = gvn().transform(new VectorCastI2XNode(iota, vt));
+  } else if (elem_bt == T_DOUBLE) {
+    iota = gvn().transform(new VectorCastL2XNode(iota, vt));
+  }
+
+  Node* index = NULL;
+  Node* scale = argument(4);
+  const TypeInt* scale_type = gvn().type(scale)->isa_int();
+  if (scale_type && scale_type->is_con() && scale_type->get_con() == 1) {
+    index = iota;
+  } else {
+    // Multiply the "scale" with "iota" if it is not "1".
+    // Broadcast the "scale" to a vector
+    switch (elem_bt) {
+      case T_BOOLEAN: // fall-through
+      case T_BYTE:    // fall-through
+      case T_SHORT:   // fall-through
+      case T_CHAR:    // fall-through
+      case T_INT: {
+        // no conversion needed
+        break;
+      }
+      case T_LONG: {
+        scale = gvn().transform(new ConvI2LNode(scale));
+        break;
+      }
+      case T_FLOAT: {
+        scale = gvn().transform(new ConvI2FNode(scale));
+        break;
+      }
+      case T_DOUBLE: {
+        scale = gvn().transform(new ConvI2LNode(scale));
+        scale = gvn().transform(new ConvL2DNode(scale));
```

Comment:
Prior target support check for these IR nodes may prevent surprises in the backend.